src/psyclone/psyGen.py

```diff
@@ -5546,15 +5558,33 @@ def module_name(self, value):
             self._module_name = value
 
     def __init__(self, name, datatype, shape=None, constant_value=None,
-                 interface=None):
+                 interface=None, precision=None):
 
         self._name = name
+        self.datatype = datatype
 
-        if datatype not in Symbol.valid_data_types:
-            raise NotImplementedError(
-                "Symbol can only be initialised with {0} datatypes but found "
-                "'{1}'.".format(str(Symbol.valid_data_types), datatype))
-        self._datatype = datatype
+        # Check that the supplied 'precision' is valid
+        if precision is not None:
+            if datatype.lower() in ["character", "boolean"]:
+                raise ValueError(
+                    "A Symbol of {0} type cannot have an associated "
+                    "precision".format(datatype.lower()))
+            if not isinstance(precision, (int, Symbol.Precision, Symbol)):
+                raise TypeError(
+                    "Symbol precision must be one of integer, Symbol.Precision"
+                    " or Symbol but got '{0}'".format(
+                        type(precision).__name__))
+            if isinstance(precision, int) and precision <= 0:
+                raise ValueError("The precision of a Symbol when specified as "
+                                 "an integer number of bytes must be > 0 but "
+                                 "got {0}".format(precision))
+            if isinstance(precision, Symbol) and \
+               (precision.datatype not in ["integer", "deferred"]
+                or precision.is_array):
+                raise ValueError("A Symbol representing the precision of "
```

Comment:
You allow "deferred here too" but do not say so in the error condition.

Reply:
Oh yes.